gammapy/scripts/analysis.py

```diff
@@ -0,0 +1,134 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Session class driving the high-level interface API"""
+import copy
+import jsonschema
+import logging
+from ..utils.scripts import read_yaml
+from astropy import units as u
+from pathlib import Path
+
+__all__ = ["Analysis"]
+
+log = logging.getLogger(__name__)
+CONFIG_PATH = Path(__file__).resolve().parent / "config"
+SCHEMA_FILE = CONFIG_PATH / "schema.yaml"
+
+
+class Analysis:
+    """The Analysis class drives an analysis working session using the high-level interface API.
+    It is initialized by default with a set of configuration parameters and values declared in
+    an internal configuration schema YAML file, though you can also provide your own configuration
+    file, as well as configuration parameters passed as a nested dictionary at the moment of
+    instantiation. In that case these parameters will overwrite the values of those present in the
+    configuration file.
+
+    Parameters
+    ----------
+    configfile : string
+        The name of a user defined configuration file.
+    config : dict
+        A nested dictionary with configuration parameters and values.
+
+    Examples
+    --------
+    Here are different examples on how to create an `Analysis` session class:
+
+    >>> from gammapy.scripts import Analysis
+    >>> cfg = {"general": {"out_folder": "myfolder"}}
+    >>> analysis = Analysis(configfile="myfile.yaml", config=cfg)
+    >>> analysis = Analysis(config=cfg)
+    >>> analysis = Analysis()
+    """
+
+    def __init__(self, configfile="", config=dict()):
```

Comment:
Avoid mutable default arguments: see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

I'm -1 to have The Analysis init have to deal with two ways to create a config.
Why not keep it simple and just take a `config`, and if someone has a config file, they write this:
```
config = Config.read("config.yaml")
analysis = Analysis(config)
```
Note that very soon Analysis will grow other things, because e.g. complex models and datasets and observations can't be created from config (maybe for now, maybe forever). So it'll really be nice if Analysis doesn't need any config-read related logic and arguments.

Overall config could have a default of an empty dict, but that should be in the Config class init, and it should use this pattern:
ttps://docs.python-guide.org/writing/gotchas/#mutable-default-arguments